### PR TITLE
New Obelisk Message Types

### PR DESCRIPTION
**Pull request by alberthli**
This PR creates generic standard message types for multiarrays and adds a `MujocoImage` message type in preparation for offscreen mujoco rendered image measurements.

### CODE REVIEW

File: obelisk_ws/src/obelisk_msgs/obelisk_sensor_msgs/CMakeLists.txt [hunk 1]
```diff
@@ -12,12 +12,15 @@ endif()
 find_package(ament_cmake REQUIRED)
 find_package(rosidl_default_generators REQUIRED)
 find_package(std_msgs REQUIRED)
+find_package(obelisk_std_msgs REQUIRED)
```
Comment (Zolkin1) [inline obelisk_ws/src/obelisk_msgs/obelisk_sensor_msgs/CMakeLists.txt:15]: by having the messages rely on other messages, building them might be a bit of a pain. maybe we should add another pixi task that builds first the "core" messages that don't rely on other messages, then builds the other messages. Then we can make `messages-build` depend on this new task so everything is built correctly.

**Review by Zolkin1 — APPROVED** [review]
lgtm. made a comment on a pixi feature that would be nice to have

File: obelisk/cpp/utils/include/msg_conversions.h [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+#include "obelisk_std_msgs/msg/u_int8_multi_array.hpp"
+
+namespace obelisk::utils::msgs {
+    namespace internal {
+        /**
+         * @brief Internal helper function to create tensors of arbitrary dimension.
+         *
+         * @param data the data to assign to the tensor
+         * @param dims the dimensions of the tensor
+         * @param Indices the index sequence to use for the parameter pack
```
Comment (alberthli) [inline obelisk/cpp/utils/include/msg_conversions.h:13]: why uppercase?

File: obelisk/cpp/utils/include/msg_conversions.h [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    } // namespace internal
+
+    /**
+     * @brief Convert MultiArray message to Eigen Tensor array.
+     *
+     * @param msg the message to convert
+     * @return the Eigen Tensor
+     */
+    template <int Size>
+    Eigen::Tensor<double, Size> MutliArrayToTensor(const obelisk_std_msgs::msg::FloatMultiArray& msg) {
```
Comment (alberthli) [inline obelisk/cpp/utils/include/msg_conversions.h:32]: sp: MultiArrayToTensor

File: obelisk/cpp/utils/include/msg_conversions.h [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    /**
+     * @brief Convert MultiArray message to Eigen Tensor array.
+     *
+     * @param msg the message to convert
+     * @return the Eigen Tensor
+     */
+    template <int Size>
+    Eigen::Tensor<uint8_t, Size> MutliArrayToTensor(const obelisk_std_msgs::msg::UInt8MultiArray& msg) {
```
Comment (alberthli) [inline obelisk/cpp/utils/include/msg_conversions.h:59]: sp: MultiArrayToTensor

File: obelisk/cpp/utils/include/msg_conversions.h [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    template <int Size>
+    obelisk_std_msgs::msg::FloatMultiArray TensorToMultiArray(const Eigen::Tensor<double, Size>& tensor) {
+        obelisk_std_msgs::msg::FloatMultiArray msg;
+        msg.layout.data_offset = 0;
+
+        // Get data into flat vector
+        msg.data.resize(tensor.size());
+        std::copy(tensor.data(), tensor.data() + tensor.size(), msg.data.begin());
+
+        // // Compute stride lengths
```
Comment (alberthli) [inline obelisk/cpp/utils/include/msg_conversions.h:94]: nit: double comment

File: obelisk/cpp/utils/include/msg_conversions.h [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    template <int Size>
+    obelisk_std_msgs::msg::UInt8MultiArray TensorToMultiArray(const Eigen::Tensor<uint8_t, Size>& tensor) {
+        obelisk_std_msgs::msg::UInt8MultiArray msg;
+        msg.layout.data_offset = 0;
+
+        // Get data into flat vector
+        msg.data.resize(tensor.size());
+        std::copy(tensor.data(), tensor.data() + tensor.size(), msg.data.begin());
+
+        // // Compute stride lengths
```
Comment (alberthli) [inline obelisk/cpp/utils/include/msg_conversions.h:128]: nit: double comment

File: tests/tests_cpp/tests_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+                        for (int n = 0; n < tensor2.dimension(4); n++) {
+                            tensor2(i, j, k, m, n) = rand() % 100;
+                        }
+                    }
+                }
+            }
+        }
+
+        auto msg                         = obelisk::utils::msgs::TensorToMultiArray<5>(tensor2);
+        Eigen::Tensor<double, 5> tensor3 = obelisk::utils::msgs::MutliArrayToTensor<5>(msg);
```
Comment (alberthli) [inline tests/tests_cpp/tests_utils.cpp:22]: make sure to fix spelling error here, too

File: tests/tests_cpp/tests_utils.cpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
+#include <catch2/catch_test_macros.hpp>
+#include <iostream>
+
+#include "msg_conversions.h"
+
+TEST_CASE("Multiarray Testing", "[utils][msgs]") {
+    for (int l = 0; l < 10; l++) {
```
Comment (alberthli) [inline tests/tests_cpp/tests_utils.cpp:7]: can you add a clarification comment here? You're running this test 10 times for an array with 5 dimensions, right?

File: obelisk/cpp/utils/include/msg_conversions.h (hunk 2 above)
Comment (Zolkin1) [reply obelisk/cpp/utils/include/msg_conversions.h:15]: It's technically a template argument, so i've named it as such

File: obelisk_ws/src/obelisk_msgs/obelisk_sensor_msgs/msg/MujocoImage.msg [hunk 9]
```diff
@@ -0,0 +1,2 @@
+string MESSAGE_NAME="MujocoImage"
```
[the remark this answers is not in the document]
Comment (alberthli) [reply obelisk_ws/src/obelisk_msgs/obelisk_sensor_msgs/msg/MujocoImage.msg]: As per slack, `MujocoImage` -> `ObkImage`

**Pull request merged**